### DISCUSSION TITLE
寫code真的要注意你寫哪個語言!

### DISCUSSION THREAD

**Discussion by henrywen547** [1101-GS4719-C/Python-Community-GS4719-C#449] [General]
我發現各種語言對於邏輯運算似乎有不同的解讀，我就單從我學過的語言來說好了。
## Python
Python的非0數字或非空字串會被當成True
如果是0或空字串，就會被當成False
```
a=2
b="Apple"
c=""
print(a or 0) #會印2
print(b and True)#會印True(因為邏輯判斷原則)
print(True and b)#會印Apple
print(c and True)#會印""
```
## C
C的int 型別能當True 或False 使用(非0或0)
C的字元陣列(字串的概念)只能當True
```
int a=2;
char b[6]="Apple";
char c[2]="";
if(a){
    printf("a");//會印a
}
if(b){
    printf("b");//會印b
}
if(!c){
    printf("c");//不會印c
}else if(c){
    printf("d");//會印d
}
```
## Java
有boolean這個東東
數字型別不能當True or False 使用
字串不能單獨當True or False 使用
```
int a=2;
String b="Apple";
boolean c=true;
if(a){
}//會錯
if(b){
}//會錯
if(b!=null){
    System.out.println("b");//沒錯，可以印出b
}
if(c){
    System.out.println("c");//沒錯
}
```
知道其他語言使用規則的大大可以幫忙補充!

**Comment by JSHT** [comment]
這邊支援 JavaScript 的邏輯運算梗圖
![image](https://user-images.githubusercontent.com/48956859/157453202-c8a4f4f8-0a75-4f76-8ef0-163174a1d86f.png)
[JavaScript 梗圖詳解](https://blog.csdn.net/weixin_46170034/article/details/107029540)

**Comment by cordydeng** [comment]
多謝大神的提醒，不過我覺得我不會遇到這種問題，因為我會的自然語言加上程式語言可能只有3種(中文.英文和python)而已哈哈

**Comment by lx3211** [comment]
謝謝整理，今天寫小測驗的時候就差點搞混了
9 or 0 -> 9而不是true

**Comment by JunTingLin** [comment]
不懂的朋友們，可以來看看這篇喔😆
[python輯運算的解讀 #463](https://github.com/NCU-GS4719-Python/Python-Community/discussions/463)

**Comment by ghost** [comment]
原本想在第一個python找規律，但發現好像又沒有

**Comment by xkllkx** [comment]
謝謝大神的提醒，之前就有遇到過這個問題。

**Comment by PoooYen** [comment]
Python為什麼越來越普及，我想就是因為最簡便最值觀吧...

**Comment by yi-1111** [comment]
認真有時真的會搞混

**Comment by d26488399** [comment]
感謝分享，真的是要多練習才不會搞混

**Comment by d26488399** [comment]
感謝分享，真的是要多練習才不會搞混

**Comment by Lai27330** [comment]
謝謝整理及分享，同時選修不同程式語言有時的確容易不小心混淆語法，之前期中時就有遇到這個問題。

**Comment by hapiwangy** [comment]
的確是要慎選語言，對於自己的專案如果用正確的語言可以'方便許多

**Comment by Ku-Hao-Chen** [comment]
謝謝分享，要多多練習才不會出錯。

**Comment by yenwei12138** [comment]
好險python是我學的第一種程式語言，目前還沒有這種問題哈，以後學習其他語言會注意。

**Comment by Rosancl** [comment]
謝謝整理！ 這學期初三個語言都有使用，有時候真的不小心會搞混qq

**Comment by yucheng1028** [comment]
謝謝整理

**Comment by Sazmafuji** [comment]
這也太有趣吧lol

**Comment by minniewei** [comment]
謝謝整理！都是我曾經困惑的地方

**Comment by Zoe06060606** [comment]
不同語言真的差超多
因為之前比較少用python寫程式
導致這學期在寫作業的時候還要google while，if的語法🥹

**Comment by harrylee1971** [comment]
每次在切換語言真的很容易打錯，學越多反而錯越多，python跟C的for迴圈型態就不一樣，每次都會手賤打到for(int i = 0)

**Comment by s96015welson** [comment]
因為每種語言都會有其特有的表示方法，所以我通常會使用最通俗易懂的表示方式，這樣在切換使用各種語言時較不易誤用，可讀性也比較高。只有當長時間寫固定一種程式語言時，為了快速撰寫或才會使用省略或簡潔一點的語法，但就會搭配詳細的註解不然隔一段時間就看不懂了。

**Comment by a45320614** [comment]
從開始學程式到現在，遇到不少這種不同程式語言間的小細節，在寫題目或考試的時候常常會不小心搞混，看來只能透過練習加深自己的印象了。

**Comment by BlahFish1006** [comment]
謝謝整理！這個真的對多語言使用者來說很困擾QQ
常常想題目都需要用不同的邏輯思路去想，
有時候有些套件python有C沒有，有時候這個用法適用C++而python不能用，
我就常常被這種問題卡很久......希望有一天也能成為轉換自如的人。
>>
這邊附上之前上課測驗時寫錯的題目：(我直接整理進筆記裡了XD)
3. 下列哪一個是Python的無窮迴圈?
    - while true: (X) ：True才是python內建語法
    - while 99:   (O) ：除了0為False，其餘非0均True

**Comment by tower20021202** [comment]
謝謝版主的整理，讓我不會再每次都搞混
